executor/read/step_executor.rs

```diff
@@ -0,0 +1,251 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{collections::HashSet, sync::Arc};
+
+use compiler::{
+    executable::{
+        function::{ExecutableFunction, ExecutableReturn},
+        match_::planner::{
+            function_plan::ExecutableFunctionRegistry,
+            match_executable::{ExecutionStep, MatchExecutable},
+        },
+        pipeline::ExecutableStage,
+    },
+    VariablePosition,
+};
+use concept::{error::ConceptReadError, thing::thing_manager::ThingManager};
+use ir::pipeline::function_signature::FunctionID;
+use itertools::Itertools;
+use storage::snapshot::ReadableSnapshot;
+
+use crate::read::{
+    collecting_stage_executor::CollectingStageExecutor,
+    immediate_executor::ImmediateExecutor,
+    nested_pattern_executor::{
+        IdentityMapper, InlinedFunctionMapper, LimitMapper, NegationMapper, NestedPatternBranch, NestedPatternExecutor,
+        OffsetMapper,
+    },
+    pattern_executor::PatternExecutor,
+};
+
+pub(super) enum StepExecutors {
+    Immediate(ImmediateExecutor),
+    Branch(NestedPatternExecutor),
+    CollectingStage(CollectingStageExecutor),
+    ReshapeForReturn(Vec<VariablePosition>),
+}
+
+impl StepExecutors {
+    pub(crate) fn unwrap_immediate(&mut self) -> &mut ImmediateExecutor {
+        match self {
+            StepExecutors::Immediate(step) => step,
+            _ => unreachable!(),
+        }
+    }
+
+    pub(crate) fn unwrap_nested_pattern_branch(&mut self) -> &mut NestedPatternExecutor {
```

Comment:
In your naming convention here, the `unwrap` is named after the enum, so this should be `unwrap_branch`!